File: java/src/main/java/Triangle.java
import java.util.Arrays;
import java.util.List;
import java.util.HashSet;

class Triangle {
    private final int uniqueSidesCount;

    Triangle(double side1, double side2, double side3) throws TriangleException {
        // All sides should be positive
        if (side1 <= 0 || side2 <= 0 || side3 <= 0) {
            throw new TriangleException();
        }
        
        // Longest side should be smaller than sum of two shortest sides
        final List<Double> sides = Arrays.asList(Double.valueOf(side1), Double.valueOf(side2), Double.valueOf(side3));
        sides.sort(Double::compareTo);
        if (sides.get(0) + sides.get(1) <= sides.get(2)) {
            throw new TriangleException();
        }
        
        // Count the sides - one of those rare moments where I don't care about generics
        this.uniqueSidesCount = new HashSet(sides).size();
    }

    boolean isEquilateral() {
        return uniqueSidesCount == 1;
    }

    boolean isIsosceles() {
        return uniqueSidesCount <= 2 ;
    }

    boolean isScalene() throws TriangleException {
        return uniqueSidesCount == 3;
    }

}

File: kotlin/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(val a: T, val b: T, val c: T) {

    init {
        val sides = listOf(a, b, c).map(Number::toDouble).sorted()
        require(sides[0] > 0 && sides[0] + sides[1] > sides[2]) {
            "Not a triangle"
        }
    }

    private val uniqueSideLengthsCount = setOf(a, b, c).size
    val isEquilateral: Boolean = uniqueSideLengthsCount == 1
    val isIsosceles: Boolean = uniqueSideLengthsCount <= 2
    val isScalene: Boolean = uniqueSideLengthsCount == 3
}
